### ISSUE TITLE
Error deploying TensorRT engine on Triton, possible nonzero op issue with data-dependent output shape.

### ISSUE THREAD

**Issue by deepmirror-wujixiu** [triton-inference-server/server#5576]
**Description**

When I attempt to use the attached onnx model to convert to a TensorRT engine, everything works fine if I use the TensorRT trtexec for building and inference. However, if I try to deploy the generated trt engine on Triton, an error is thrown with the message "unexpected size for output 'output', byte-size 0 is less than 1 * 3053044".

Upon further investigation, I discovered that the error occurs at line 1108 in instance_state.cc in the tensorrt_backend repository. The model has two nonzero ops, and I believe that the main error is caused by the nonzero op.

In light of this, I am uncertain about how to configure the config.pbtxt if the output's shape is data-dependent. Could you please advise me on this matter?

**Triton Information**
What version of Triton are you using?

**22.12**

Are you using the Triton container or did you build it yourself?

**No**

**To Reproduce**
Steps to reproduce the behavior.
1. Convert the onnx model to tensorrt engine use trtexec
trtexec --onnx=nonzero_bug.onnx --saveEngine=model.plan
2. Config the triton model
```
platform: "tensorrt_plan"
max_batch_size: 0
input [
  {
    name: "input"
    data_type: TYPE_FP32
    format: FORMAT_NCHW
    dims: [1, 720, 1080]
  }
]
output [
  {
    name: "output"
    data_type: TYPE_INT32
    dims: [-1]
  }
]
```
3. Request inference in triton

Describe the models (framework, inputs, outputs), ideally include the model configuration file (if using an ensemble include the model configuration file for that as well).

**Expected behavior**

Get nonzero index of the input.
[nonzero_bug.zip](https://github.com/triton-inference-server/server/files/11121072/nonzero_bug.zip)

**Comment by dyastremsky** [comment]
I don't believe we have a mechanism in Triton for allowing variable-sized output where the shape is determined at inference time. Tagging @tanmayv25 and @GuanLuo who may know more about this use case.

**Comment by tanmayv25** [comment]
We do support data-dependent output shape in Triton. We also have a model with non_zero layer in our CI testing
Looking at the Triton error, it seems [getMaxOutputSize()](https://github.com/triton-inference-server/tensorrt_backend/blob/f5bf02b4e25c385dd0a03288fbb845cf933540d5/src/instance_state.cc#L3978) returned 0 as output which caused [this](https://github.com/triton-inference-server/tensorrt_backend/blob/main/src/instance_state.cc#L1102) check to fail.

There are some bugs in TensorRT library where [getMaxOutputSize](https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/classnvinfer1_1_1_i_execution_context.html#a9567a15348ca73d1e8c371c23acd3c5d) TensorRT function returns 0 instead of an upper bound for the model (or -1 if it fails to obtain such upper-bound).

@dyastremsky  Can you open a jira bug for this issue? We can probably fix it by moving from getMaxOutputSize to OutputAllocator?

**Comment by dyastremsky** [comment]
Thanks for clarifying, Tanmay. Created a ticket to investigate.

**Comment by deepmirror-wujixiu** [comment]
@tanmayv25 Any update on this issue?

**Comment by tanmayv25** [comment]
@deepmirror-wujixiu We haven't yet prioritized this issue. We do have a ticket to move to OutputAllocator APIs which should resolve this problem.
